### PR TITLE
Access to dynamic Reusable part.

### LINKED ISSUE #4161: Dynamically created reusable parts not accessible by name
I was trying to inject a property in a Reusable Part shape but I was not able to access it using the `Shape.Named("")` method. 

Example:
- Create a Part called MyReusablePart from the admin UI.
- Attach the part to a `MyContentType` ContentType twice : "RP1", "RP2" 
 - Try to access the RP2 shape by using Model.Content.Named("RP2") does not work.

### PR DESCRIPTION
**Pull request by jtkech**
Fixes #4161, so that you can use `Model.Content.Named("Reusable1")`.

- It was working if the reusable part is created by code but not if created through the admin.

- When created by code the differentiator is set to `partName` or `{partName}-{shapeType}` but when created dynamically it was `partTypeName` (here the shapeType being `ContentPart`).

- So here, for a dynamic part i set the differentiator to `partName`.

- Then i added some alternates, e.g based on the `contentType` => `Article-Reusable1`.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs [hunk 1]
```diff
@@ -96,12 +97,12 @@ public async Task BuildDisplayAsync(ContentItem contentItem, BuildDisplayContext
...
 
-                    if (part.GetType() == typeof(ContentPart) && contentTypePartDefinition.PartDefinition.Name != contentTypePartDefinition.ContentTypeDefinition.Name)
+                    if (part.GetType() == typeof(ContentPart) && partTypeName != contentTypePartDefinition.ContentTypeDefinition.Name)
                     {
                         var shapeType = context.DisplayType != "Detail" ? "ContentPart_" + context.DisplayType : "ContentPart";
 
                         var shapeResult = new ShapeResult(shapeType, ctx => ctx.ShapeFactory.CreateAsync(shapeType, () => Task.FromResult<IShape>(new ZoneHolding(() => ctx.ShapeFactory.CreateAsync("Zone", Arguments.Empty)))));
-                        shapeResult.Differentiator(contentTypePartDefinition.PartDefinition.Name);
+                        shapeResult.Differentiator(partName);
                         shapeResult.Location("Content");
```
Comment (deanmarcussen) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs]: Looks good @jtkech 
For info @jptissot this `shapeResult.Location("Content");` is why it's also showing up in `SummaryAdmin` - this will make it show everywhere, always. For laters when we know what is the best idea for hiding this, we could do another pr to match whatever is decided.

Comment (sebastienros) [reply src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs]: Do you mean we should do a `shapeResult.Location("DetailsAdmin", "-");` too here?

Comment (jtkech) [reply src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs]: I think that's ok here, a dynamic part is composed of fields, so the rendering depends on the field drivers where most of them are not rendered in `SummaryAdmin`.

E.g if a blog post has a dyn part, when displaying the blog in `DetailAdmin`, the related list part will be rendered in `DetailAdmin` but each blogpost will be rendered in `SummaryAdmin`.

The "problem" is when the blog itself has an additional field / part (dynamic or not), it will be rendered when displaying the blog in `DetailAdmin`. But i think that there is nothing to do here, this seems to be only due to the field drivers.

Most of the field drivers already do

    .Location("SummaryAdmin", "")

Maybe we could add

    .Location("DetailAdmin", "")
